chalk-ir/src/zip.rs

```diff
@@ -21,52 +21,91 @@ use std::sync::Arc;
 /// represented by two distinct `ItemId` values, and the impl for
 /// `ItemId` requires that all `ItemId` in the two zipped values match
 /// up.
-pub trait Zipper<'i, I: Interner> {
+pub trait Zipper<'i, I: Interner + 'i> {
     /// Indicates that the two types `a` and `b` were found in matching spots.
-    fn zip_tys(&mut self, a: &Ty<I>, b: &Ty<I>) -> Fallible<()>;
+    fn zip_tys(&mut self, variance: Variance, a: &Ty<I>, b: &Ty<I>) -> Fallible<()>;
 
     /// Indicates that the two lifetimes `a` and `b` were found in matching spots.
-    fn zip_lifetimes(&mut self, a: &Lifetime<I>, b: &Lifetime<I>) -> Fallible<()>;
+    fn zip_lifetimes(
+        &mut self,
+        variance: Variance,
+        a: &Lifetime<I>,
+        b: &Lifetime<I>,
+    ) -> Fallible<()>;
 
     /// Indicates that the two consts `a` and `b` were found in matching spots.
-    fn zip_consts(&mut self, a: &Const<I>, b: &Const<I>) -> Fallible<()>;
+    fn zip_consts(&mut self, variance: Variance, a: &Const<I>, b: &Const<I>) -> Fallible<()>;
 
     /// Zips two values appearing beneath binders.
-    fn zip_binders<T>(&mut self, a: &Binders<T>, b: &Binders<T>) -> Fallible<()>
+    fn zip_binders<T>(
+        &mut self,
+        variance: Variance,
+        a: &Binders<T>,
+        b: &Binders<T>,
+    ) -> Fallible<()>
     where
         T: HasInterner<Interner = I> + Zip<I> + Fold<I, I, Result = T>;
 
-    /// Retreives the interner from the underlying zipper object
+    /// Zips two substs
+    fn zip_substs(
+        &mut self,
+        variances: Option<Vec<Variance>>,
+        a: &[GenericArg<I>],
+        b: &[GenericArg<I>],
+    ) -> Fallible<()>
+    where
+        Self: Sized,
+    {
+        for (i, (a, b)) in a.iter().zip(b.iter()).enumerate() {
+            let variance = variances.as_ref().map_or(Variance::Invariant, |v| v[i]);
```

Comment:
```suggestion
            let variance = variances.as_ref().map(|v| v[i]).unwrap_or(Variance::Invariant);
```

ymmv but I personally find this formulation much easier to read than `map_or`

Reply:
I can change this, but I actually just copied this bit from rustc.

Reply:
Doesn't matter too much. I just have a thing against `map_or` :P